vowpalwabbit/core/src/reductions/cb/cb_explore.cc

```diff
@@ -362,61 +375,86 @@ base_learner* VW::reductions::cb_explore_setup(VW::setup_base_i& stack_builder)
   if (options.was_supplied("indexing"))
   { THROW("--indexing is not compatible with contextual bandits, please remove this option") }
 
-  data->_random_state = all.get_random_state();
-  uint32_t num_actions = data->cbcs.num_actions;
-
   // If neither cb nor cats_tree are present on the reduction stack then
   // add cb to the reduction stack as the default reduction for cb_explore.
   if (!options.was_supplied("cats_tree") && !options.was_supplied("cb"))
   {
     // none of the relevant options are set, default to cb
     std::stringstream ss;
-    ss << data->cbcs.num_actions;
+    ss << cb_explore_opts->num_actions;
     options.insert("cb", ss.str());
   }
+  cb_explore_opts->cover_supplied = options.was_supplied("cover");
+  cb_explore_opts->epsilon_supplied = options.was_supplied("epsilon");
+  cb_explore_opts->bag_supplied = options.was_supplied("bag");
+  cb_explore_opts->first_supplied = options.was_supplied("first");
+  return cb_explore_opts;
+}
 
-  if (data->epsilon < 0.0 || data->epsilon > 1.0) { THROW("The value of epsilon must be in [0,1]"); }
+}  // namespace CB_EXPLORE
 
-  data->cbcs.cb_type = VW::cb_type_t::dr;
-  data->model_file_version = all.model_file_ver;
+base_learner* VW::reductions::cb_explore_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+  auto cb_explore_opts = CB_EXPLORE::get_cb_explore_options_instance(all, options);
+  if (cb_explore_opts == nullptr) { return nullptr; }
+  auto cb_explore_data = VW::make_unique<cb_explore>(all.logger);
+  cb_explore_data->cbcs.num_actions = cb_explore_opts->num_actions;
+  cb_explore_data->tau = cb_explore_opts->tau;
+  cb_explore_data->epsilon = cb_explore_opts->epsilon;
+  cb_explore_data->bag_size = cb_explore_opts->bag_size;
+  cb_explore_data->cover_size = cb_explore_opts->cover_size;
+  cb_explore_data->nounif = cb_explore_opts->nounif;
+  cb_explore_data->psi = cb_explore_opts->psi;
+
+  cb_explore_data->_random_state = all.get_random_state();
+  uint32_t num_actions = cb_explore_data->cbcs.num_actions;
+
+  if (cb_explore_data->epsilon < 0.0 || cb_explore_data->epsilon > 1.0)
```

Comment:
Should this be in get_cb_explore_options_instance?

Reply:
yea good call